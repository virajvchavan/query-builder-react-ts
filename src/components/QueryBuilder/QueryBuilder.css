.querySelector {
    width: 635px;
    padding: 10px;
}

.querySelector .allQueries {
    display: flex;
    padding: 5px;
}

.querySelector .allQueries .where {
    width: 60px;
}

.row {
    display: flex;
    margin-bottom: 18px;
}

.row>* {
    margin: 0 6px;
}

.applyBtn, .saveBtn {
    float: right;
}

.btn {
    cursor: pointer;
    padding: 10px;
    width: fit-content;
}

.removeBtn {
    line-height: 15px;
    color: red;
    font-weight: bold;
}

.addBtn,
.applyBtn,
.saveBtn ,
.clearAll {
    background-color: teal;
    color: white;
    font-weight: bold;
    border-radius: 5px;
}

.saveBtn.disabled {
    color: lightgrey;
}

.applyBtn {
    background-color: rgb(14, 123, 42);
    margin-left: 10px;
}

.saveBtn {
    margin-left: 3px;
}
.clearAll {
    background-color: #cb0000;
    margin-left: 10px;
}
.lhs {
    width: 150px;
}

.operator {
    width: 160px;
}

.rhs {
    width: 251px;
}
.queryNameInputContainer {
    display: inline-flex;
    float: right;
}
