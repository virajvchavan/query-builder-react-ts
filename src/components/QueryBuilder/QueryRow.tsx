import React, { useEffect, useState } from 'react';
import { Row } from '../../queryConfig/queryConfig';
import Select, { ValueType } from 'react-select';
import RhsInput from './RhsInput/RhsInput';
import { RhsType } from './QueryBuilder';

interface Props {
    lhs: string,
    operator: string,
    rhs?: string | Array<number> | Array<string>,
    index: number,
    removeRow: (index: number) => void,
    queryConfig: Row,
    lhsOptions: Array<OptionType>,
    onLhsChange: (index: number, lhs: ValueType<OptionType>) => void,
    onOperatorChange: (index: number, operator: ValueType<OptionType>) => void,
    onRhsChange: (index: number, rhs: RhsType) => void,
    onCustomSelectRhsChange: Function
}

export type OptionType = {
    value: string;
    label: string | number;
};

export type JsonType = {
    [key: string]: string | number
}

export default function QueryRow({ lhs, rhs, operator, index, queryConfig, removeRow, lhsOptions, onLhsChange, onOperatorChange, onRhsChange, onCustomSelectRhsChange }: Props) {
    const [operatorOptions, setOperatorOptions] = useState<Array<{ value: string, label: string }>>();
    const [operatorValue, setOperatorValue] = useState<OptionType>();

    useEffect(() => {
        let operatorOptions: Array<{ value: string, label: string }> = [];
        queryConfig.operators.forEach((operator, index) => {
            operatorOptions.push({ value: operator.value, label: operator.text });
        });
    
        setOperatorOptions(operatorOptions);
        setOperatorValue({
            label: operatorOptions?.find(item => item.value === operator)?.label || operator,
            value: operator
        });
    }, [operator, queryConfig]);

    const onRemoveClick = () => removeRow(index);

    const onNormalRhsChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        onRhsChange(index, evt.target.value);
    }

    const onSelectRhsChange = (newElement: ValueType<OptionType>) => {
        onCustomSelectRhsChange(index, newElement)
    }

    const onMultiSelectNumbersRhsChange = (newRhs: Array<number>) => {
        onRhsChange(index, newRhs);
    }

    let lhsValue: OptionType = {
        label: queryConfig.label,
        value: lhs
    };

    return <div className="row" data-testid="queryRow">
        <div className="lhs">
            <Select
                options={lhsOptions}
                onChange={(newValue: ValueType<OptionType>) => onLhsChange(index, newValue)}
                defaultValue={lhsOptions[0]} value={lhsValue}
            />
        </div>
        <div className="operator">
            <Select
                options={operatorOptions}
                onChange={(newValue: ValueType<OptionType>) => onOperatorChange(index, newValue)}
                value={operatorValue}
            />
        </div>
        <div className="rhs">
            <RhsInput rhs={rhs} queryConfig={queryConfig} onNormalRhsChange={onNormalRhsChange} onSelectRhsChange={onSelectRhsChange} onMultiSelectNumbersRhsChange={onMultiSelectNumbersRhsChange} />
        </div>
        <div className="btn removeBtn" data-testid="removeRowBtn" onClick={onRemoveClick}>X</div>
    </div>;
}
