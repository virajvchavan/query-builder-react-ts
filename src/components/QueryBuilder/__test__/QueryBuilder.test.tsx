import { cleanup, render } from '@testing-library/react';
import React from "react";
import userEvent from '@testing-library/user-event'
import QueryBuilder from "../QueryBuilder";
import queryConfig from '../../../queryConfig/queryConfig';

// Mostly integration tests

afterEach(cleanup)

test("it should have mandatory elements", () => {
    const { getByText } = render(<QueryBuilder queryConfig={queryConfig} />);

    expect(getByText("+ Add")).not.toBeNull();
    expect(getByText("Apply")).not.toBeNull();
});

test("add-button click should add a new row", () => {
    const { getByText, queryAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    expect(queryAllByTestId("queryRow").length).toBe(1);
    userEvent.click(getByText("+ Add"));
    expect(queryAllByTestId("queryRow").length).toBe(2);
    userEvent.click(getByText("+ Add"));
    expect(queryAllByTestId("queryRow").length).toBe(3);
});

test("remove-row-button click should remove a row", () => {
    const { getByText, queryAllByTestId, getAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    userEvent.click(getByText("+ Add"));
    userEvent.click(getByText("+ Add"));
    userEvent.click(getAllByTestId("removeRowBtn")[0]);
    expect(queryAllByTestId("queryRow").length).toBe(2);
});

test("it should remove all rows when Clear All button is clicked", () => {
    const { getByText, queryAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    userEvent.click(getByText("+ Add"));
    userEvent.click(getByText("+ Add"));
    userEvent.click(getByText("Clear"));
    expect(queryAllByTestId("queryRow").length).toBe(0);
});

test("it should show notice when no rows are added", () => {
    const { getByText, getAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    userEvent.click(getAllByTestId("removeRowBtn")[0]);
    expect(getByText("Click on `+ Add` to add a query.")).not.toBeNull();
});

test("it should render correct LHS options", () => {
    const { getByText, queryAllByText } = render(<QueryBuilder queryConfig={queryConfig} />);
    userEvent.click(getByText(queryConfig[Object.keys(queryConfig)[0]].label)); // this opens up the dropdown

    Object.keys(queryConfig).forEach(key => {
        expect(queryAllByText(queryConfig[key].label)[0]).toBeInTheDocument();
    });
});

test("correct operator list is rendered based on LHS", () => {
    const { queryAllByText } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.operators[0].text)[0]); // this opens up the dropdown
    rowConfig1.operators.forEach(operator => {
        expect(queryAllByText(operator.text)[0]).toBeInTheDocument();
    });

    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    let rowConfig2 = queryConfig[Object.keys(queryConfig)[1]];
    userEvent.click(queryAllByText(rowConfig2.label)[0]);

    userEvent.click(queryAllByText(rowConfig2.operators[0].text)[0]);
    rowConfig2.operators.forEach(operator => {
        expect(queryAllByText(operator.text)[0]).toBeInTheDocument();
    });
});

test("correct RHS element is rendered when RHS type in the provided config is text", () => {
    const { queryAllByText, getByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Campaign")[0]);

    let rhsElement = getByTestId("rhsInput");
    expect(rhsElement.getAttribute("type")).toBe("text");
});

test("correct RHS element is rendered when RHS type in the provided config is number", () => {
    const { queryAllByText, getByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Revenue")[0]);

    let rhsElement = getByTestId("rhsInput");
    expect(rhsElement.getAttribute("type")).toBe("number");
});

test("correct RHS element is rendered when RHS type in the provided config is multi-select-numbers-in-a-range", () => {
    const { queryAllByText, getByLabelText } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Account")[0]);

    let rhsElement = getByLabelText("Enter values");
    expect(rhsElement.getAttribute("type")).toBe("number");
});

test("correct RHS element is rendered when RHS type in the provided config is multi-select-list", () => {
    const { getByText, queryAllByText } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Country")[0]);

    expect(getByText("Select...")).not.toBeNull();
});

test("it should save a query with a name", () => {
    window.localStorage.clear();
    const { getByTestId, queryAllByText, getByText, getByPlaceholderText } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Campaign")[0]);

    userEvent.type(getByTestId("rhsInput"), "bangalore");

    let queryNameInput = getByPlaceholderText("Name your query");
    userEvent.type(queryNameInput, "query1");
    userEvent.click(getByText("Save for later"));

    // value should clear after saving the query
    expect(queryNameInput).toHaveValue("");

    // check if the newly created query is in the savedQueries dropdown
    userEvent.click(getByText("Select to Apply"));
    expect(getByText("query1")).not.toBeNull();
    cleanup();
});

test("it should apply a saved query correctly", () => {
    window.localStorage.clear();
    const { getByTestId, queryAllByText, getByText, getByPlaceholderText, queryAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Campaign")[0]);

    userEvent.type(getByTestId("rhsInput"), "bangalore");

    let queryNameInput = getByPlaceholderText("Name your query");
    userEvent.type(queryNameInput, "query1");
    userEvent.click(getByText("Save for later"));

    // value should clear after saving the query
    expect(queryNameInput).toHaveValue("");

    // now remove all rows
    userEvent.click(getByText("Clear"));

    // select the newly created query from savedQueries dropdown
    userEvent.click(getByText("Select to Apply"));
    expect(getByText("query1")).not.toBeNull();
    userEvent.click(getByText("query1"));

    // now the rows should be added from the saved query
    expect(queryAllByTestId("queryRow").length).toBe(1);
    expect(getByTestId("rhsInput")).toHaveValue("bangalore");
});

test("it should call console.log with correct json when apply btn is clicked", () => {
    console.log = jest.fn();

    const { getByTestId, queryAllByText, getByText } = render(<QueryBuilder queryConfig={queryConfig} />);
    let rowConfig1 = queryConfig[Object.keys(queryConfig)[0]];
    userEvent.click(queryAllByText(rowConfig1.label)[0]);
    userEvent.click(queryAllByText("Campaign")[0]);

    userEvent.type(getByTestId("rhsInput"), "bangalore");

    userEvent.click(getByText("Apply"));

    let rowConfig = queryConfig["campaign"];
    expect(console.log).toHaveBeenCalledWith([{
        lhs: "campaign",
        operator: rowConfig.operators[0].value,
        rhs: "bangalore"
    }]);
});

test("it should match snapshot for the first render", () => {
    const { asFragment } = render(<QueryBuilder queryConfig={queryConfig} />)
    expect(asFragment()).toMatchSnapshot()
});

test("it should match snapshot when a new row is added", () => {
    const { asFragment, getByText, queryAllByTestId } = render(<QueryBuilder queryConfig={queryConfig} />);
    expect(queryAllByTestId("queryRow").length).toBe(1);
    userEvent.click(getByText("+ Add"));
    expect(asFragment()).toMatchSnapshot();
});
