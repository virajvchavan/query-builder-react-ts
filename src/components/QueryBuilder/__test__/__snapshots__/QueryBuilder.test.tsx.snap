// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match snapshot for the first render 1`] = `
<DocumentFragment>
  <div
    class="querySelector"
  >
    <div
      class="savedQueries"
    >
      <label
        class="savedQueries-title"
        for="savedQueries-input"
      >
        Saved filters: 
      </label>
      <div
        class="savedQueries-inputContainer"
      >
        <div
          class=" css-2b097c-container"
          id="savedQueries-input"
        >
          <div
            class=" css-yk16xz-control"
          >
            <div
              class=" css-g1d714-ValueContainer"
            >
              <div
                class=" css-1wa3eu0-placeholder"
              >
                Select to Apply
              </div>
              <div
                class="css-b8ldur-Input"
              >
                <div
                  class=""
                  style="display: inline-block;"
                >
                  <input
                    aria-autocomplete="list"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    id="react-select-60-input"
                    spellcheck="false"
                    style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-6q0nyr-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="allQueries"
    >
      <div
        class="where"
      >
        where
      </div>
      <div
        class="queryRows"
      >
        <div
          class="row"
          data-testid="queryRow"
        >
          <div
            class="lhs"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    Account
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-61-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="operator"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    contains
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-62-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="rhs"
          >
            <div
              class="multi-input-container"
            >
              <div
                class="multi-input-control"
              >
                <div
                  class="multi-select-valueContainer"
                >
                  <div
                    class="multi-select-inputBox"
                  >
                    <div
                      style="display: inline-block; width: 100%;"
                    >
                      <label
                        for="multi-value-input"
                        style="display: none;"
                      >
                        Enter values
                      </label>
                      <input
                        id="multi-value-input"
                        max="1000"
                        min="1"
                        placeholder="Enter values (1-1000)"
                        type="number"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="multi-select-BtnContainer"
                >
                  <div
                    class="multi-select-btnSeparator"
                  />
                  <div
                    class="multi-select-btn add-all-btn"
                  >
                    +
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn removeBtn"
            data-testid="removeRowBtn"
          >
            X
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn addBtn"
    >
      + Add
    </button>
    <button
      class="btn clearAll"
    >
      Clear
    </button>
    <button
      class="btn applyBtn"
    >
      Apply
    </button>
    <button
      class="btn saveBtn disabled"
    >
      Save for later
    </button>
    <div
      class="queryNameInputContainer"
    >
      <input
        class="styled-input"
        placeholder="Name your query"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`it should match snapshot when a new row is added 1`] = `
<DocumentFragment>
  <div
    class="querySelector"
  >
    <div
      class="savedQueries"
    >
      <label
        class="savedQueries-title"
        for="savedQueries-input"
      >
        Saved filters: 
      </label>
      <div
        class="savedQueries-inputContainer"
      >
        <div
          class=" css-2b097c-container"
          id="savedQueries-input"
        >
          <div
            class=" css-yk16xz-control"
          >
            <div
              class=" css-g1d714-ValueContainer"
            >
              <div
                class=" css-1wa3eu0-placeholder"
              >
                Select to Apply
              </div>
              <div
                class="css-b8ldur-Input"
              >
                <div
                  class=""
                  style="display: inline-block;"
                >
                  <input
                    aria-autocomplete="list"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    id="react-select-63-input"
                    spellcheck="false"
                    style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-6q0nyr-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="allQueries"
    >
      <div
        class="where"
      >
        where
      </div>
      <div
        class="queryRows"
      >
        <div
          class="row"
          data-testid="queryRow"
        >
          <div
            class="lhs"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    Account
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-64-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="operator"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    contains
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-65-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="rhs"
          >
            <div
              class="multi-input-container"
            >
              <div
                class="multi-input-control"
              >
                <div
                  class="multi-select-valueContainer"
                >
                  <div
                    class="multi-select-inputBox"
                  >
                    <div
                      style="display: inline-block; width: 100%;"
                    >
                      <label
                        for="multi-value-input"
                        style="display: none;"
                      >
                        Enter values
                      </label>
                      <input
                        id="multi-value-input"
                        max="1000"
                        min="1"
                        placeholder="Enter values (1-1000)"
                        type="number"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="multi-select-BtnContainer"
                >
                  <div
                    class="multi-select-btnSeparator"
                  />
                  <div
                    class="multi-select-btn add-all-btn"
                  >
                    +
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn removeBtn"
            data-testid="removeRowBtn"
          >
            X
          </div>
        </div>
        <div
          class="row"
          data-testid="queryRow"
        >
          <div
            class="lhs"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    Account
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-66-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="operator"
          >
            <div
              class=" css-2b097c-container"
            >
              <div
                class=" css-yk16xz-control"
              >
                <div
                  class=" css-g1d714-ValueContainer"
                >
                  <div
                    class=" css-1uccc91-singleValue"
                  >
                    contains
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class=""
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-67-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="rhs"
          >
            <div
              class="multi-input-container"
            >
              <div
                class="multi-input-control"
              >
                <div
                  class="multi-select-valueContainer"
                >
                  <div
                    class="multi-select-inputBox"
                  >
                    <div
                      style="display: inline-block; width: 100%;"
                    >
                      <label
                        for="multi-value-input"
                        style="display: none;"
                      >
                        Enter values
                      </label>
                      <input
                        id="multi-value-input"
                        max="1000"
                        min="1"
                        placeholder="Enter values (1-1000)"
                        type="number"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="multi-select-BtnContainer"
                >
                  <div
                    class="multi-select-btnSeparator"
                  />
                  <div
                    class="multi-select-btn add-all-btn"
                  >
                    +
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn removeBtn"
            data-testid="removeRowBtn"
          >
            X
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn addBtn"
    >
      + Add
    </button>
    <button
      class="btn clearAll"
    >
      Clear
    </button>
    <button
      class="btn applyBtn"
    >
      Apply
    </button>
    <button
      class="btn saveBtn disabled"
    >
      Save for later
    </button>
    <div
      class="queryNameInputContainer"
    >
      <input
        class="styled-input"
        placeholder="Name your query"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
